### ISSUE TITLE
Migrate build to gradle 

### ISSUE THREAD

**Issue by ieugen** [jbake-org/jbake#276]
There is a desire to migrate the project to gradle for a better develoment experience and build capabilities
- easy multi module management
- easy release artifact building
- easy releasing and versioning (gradle-git plugins)

Sub-tasks:
- merge existing PR-s
- migrate the code base 
- make release with new process
- docuemnt the release process

**Comment by ieugen** [comment]
After #275

**Comment by ancho** [comment]
:+1:

**Comment by jangalinski** [comment]
This has all potential to become a religious discussion, but could someone please summarize the benefits you see in changing the build system? In my experience, building dev communities around the most common tools is more successful ...and that's maven imho.

**Comment by ancho** [comment]
@jangalinski There are plenty of benefits switching to gradle as a build management tool.
In my opinion it is the build tool of the present and the future. Many big companies and communities are switching to gradle or already did.

* It's fast
* It's flexible
* It's easy to learn
* has a powerful DSL with convention over configuration paradigm
* a lot of plugins for nearly everything
* a big and nice community
* very good documentation
* it's fun

I would suggest you follow a few links and build your own opinion:

* [why gradle?](http://gradle.org/whygradle-build-automation/)
* [maven vs. gradle](http://gradle.org/maven_vs_gradle)
* [plugins](https://plugins.gradle.org/)

The best thing is to try it yourself:

* Install [sdkman](http://sdkman.io/) to install the latest version of gradle
  
```
  curl -s get.sdkman.io | bash
  sdk install gradle
```
* create your first java gradle project

```
   mkdir /tmp/gradleTest
   cd /tmp/gradleTest
   gradle init --type java-library
```

* run the build

```
  gradle build
```

* Have a look at build.gradle file to see the configuration that was generated by the init tasks

Moving to gradle was a very good decision for me. I never regretted it and I don't want to switch back.
Just my few cents :)

**Comment by jangalinski** [comment]
Thanks @ancho for the links. I already work with some projects using gradle and it works well. Its just a different command  line call. When setting up my own stuff, I stick to maven so far, its what me and my team know best and I don't have to think about it. I would be open to learn more about gradle ...

I just question that it is worth migrating. More "like: If you do not have a build system and your developers know/are open to gradle: use gradle" ... "if you are already on maven and encounter no problems: stay on maven"

I lack to see the specific benefits that jbake would have by switching the build system. What of the impressive "gradle can, maven doesn't" list is crucial/missed for the success of jbake?

**Comment by ieugen** [comment]
@ancho very good arguments, thank you.

@jangalinski : I've used Maven a lot and it had made my life easier. Having discovered gradle for the ~2 years, over time, I've switched all of my builds to it and never look back. 

One of the biggest reasone is the flexibility and plugins. You can build binary distributions very easy and handle the project life cycle much better than with maven, especially in poliglot multi-module builds. 

Gradle makes impossible things in Maven possible. It's build happiness :) .

A few cons: it is a bit behind in IDE integration and you might hit some issues that you have to fix, but once fixed, things are ok.

**Comment by ieugen** [comment]
The immediate benefits would come from 

- better handling of multi-module builds (caches build results)
- more flexible binary distribution. For a sample, chek out jbakery binary distribtuion: source is [1] and binary release is [2]

[1] https://github.com/netdava/jbakery
[2] https://bintray.com/netdava/maven/jbakery

**Comment by ancho** [comment]
btw. I started to migrate jbake to gradle. It is still work in progress but works for me so far.
@ieugen do you mind to have a look at it? I could open a WIP pull request to start a discussion. What do you thin?

https://github.com/ancho/jbake/tree/feature/gradle-build

Based on that branch I started to modularize the build to get a clean separation of jbake-core and distribution packaging.

https://github.com/ancho/jbake/tree/feature/distribution-submodule

There are open questions about binary distribution. So far the gradle install task does a local maven publishing.

**Comment by ieugen** [comment]
Cool. Please do open a PR. 

I was waiting for a a response from @jonbullock on the migration status. There are curently a lot of open PR's that will need to be merged or we will lose them. That will not be so good. 

@ancho If you can wait a few more days so we can settle this, it would be great. I think I can answer most of the gradle questions if you supply them. Especially if there is code.

**Comment by ancho** [comment]
Yes. I'm aware about that. But I couldn't resist to begin the migration steps :)
The open PR's is one of the reasons I opened a new branch for the modularization part. It is more a proof of concept.
The feature/gradle-build branch should make it possible to merge all the open PR's as the basic project stucture has not changed, except fixture test resources. But that is another story.
I'll open a pull request and try to describe es much as possible.
No hurry. We have enough time.

**Comment by jonbullock** [comment]
One additional reason to switch to Gradle #189

**Comment by jonbullock** [comment]
I'm going to focus on dealing with the existing PR's and releasing v2.5.0 short term.
